File: coordinator/api/earnings_market.go
package api

import (
	"encoding/json"
	"fmt"
	"math"
	"net/http"
	"time"

	"github.com/eigeninference/d-inference/coordinator/payments/baserewards"
	"github.com/eigeninference/d-inference/coordinator/registry"
	"github.com/eigeninference/d-inference/coordinator/store"
)

const earningsMarketWindow = 30 * 24 * time.Hour

type earningsMarketModel struct {
	ID                          string  `json:"id"`
	DisplayName                 string  `json:"display_name"`
	MinRAMGB                    int     `json:"min_ram_gb"`
	SizeBytes                   int64   `json:"size_bytes"`
	SizeGB                      float64 `json:"size_gb"`
	WorkPayoutMicroUSD          int64   `json:"work_payout_micro_usd"`
	PaidTokens                  int64   `json:"paid_tokens"`
	PaidJobs                    int64   `json:"paid_jobs"`
	AggregateTPS                float64 `json:"aggregate_tps"`
	AggregateMemoryBandwidthGBs float64 `json:"aggregate_memory_bandwidth_gbps"`
	BenchmarkTPS                float64 `json:"benchmark_tps"`
	BenchmarkMemoryBandwidthGBs float64 `json:"benchmark_memory_bandwidth_gbps"`
	ProviderSupply              int     `json:"provider_supply"`
	EstimateAvailable           bool    `json:"estimate_available"`
	UnavailableReason           string  `json:"unavailable_reason,omitempty"`
}

type earningsMarketAudit struct {
	TotalSettledWorkMicroUSD int64 `json:"total_settled_work_micro_usd"`
	ModeledWorkMicroUSD      int64 `json:"modeled_work_micro_usd"`
	UnattributedWorkMicroUSD int64 `json:"unattributed_work_micro_usd"`
	TotalPaidTokens          int64 `json:"total_paid_tokens"`
	ModeledPaidTokens        int64 `json:"modeled_paid_tokens"`
	UnattributedPaidTokens   int64 `json:"unattributed_paid_tokens"`
	TotalPaidJobs            int64 `json:"total_paid_jobs"`
	ModeledPaidJobs          int64 `json:"modeled_paid_jobs"`
	UnattributedPaidJobs     int64 `json:"unattributed_paid_jobs"`
}

type earningsMarketBaseRewards struct {
	Enabled             bool               `json:"enabled"`
	MonthlyPoolMicroUSD int64              `json:"monthly_pool_micro_usd"`
	MinUptimeFraction   float64            `json:"min_uptime_fraction"`
	ReductionK          float64            `json:"reduction_k"`
	AccountCapFraction  float64            `json:"account_cap_fraction"`
	Tiers               []baserewards.Tier `json:"tiers"`
}

type earningsMarketResponse struct {
	WindowStart time.Time                 `json:"window_start"`
	WindowEnd   time.Time                 `json:"window_end"`
	WindowDays  int                       `json:"window_days"`
	Models      []earningsMarketModel     `json:"models"`
	Audit       earningsMarketAudit       `json:"audit"`
	BaseRewards earningsMarketBaseRewards `json:"base_rewards"`
}

// handleEarningsMarket serves the one calculator-ready market snapshot used by
// both the console and static landing page. Work demand comes only from settled
// provider earnings; capacity comes only from the live routable registry.
func (s *Server) handleEarningsMarket(w http.ResponseWriter, _ *http.Request) {
	const cacheKey = "earnings_market:v1"
	if cached, ok := s.readCache.Get(cacheKey); ok {
		writeCachedJSON(w, cached)
		return
	}

	windowEnd := time.Now().UTC()
	windowStart := windowEnd.Add(-earningsMarketWindow)

	records, err := s.store.ListActiveModelRegistryWithError()
	if err != nil {
		s.writeEarningsMarketFailure(w, "list active model registry", err)
		return
	}
	aliases, err := s.store.ListModelAliases()
	if err != nil {
		s.writeEarningsMarketFailure(w, "list model aliases", err)
		return
	}
	work, err := s.store.ModelSettledWorkTotals(windowStart, windowEnd)
	if err != nil {
		s.writeEarningsMarketFailure(w, "aggregate settled work", err)
		return
	}

	response, err := buildEarningsMarketResponse(
		records,
		aliases,
		s.registry.ModelCapacitySnapshot(),
		work,
		windowStart,
		windowEnd,
		s.baseRewardsConfig,
	)
	if err != nil {
		s.writeEarningsMarketFailure(w, "build market snapshot", err)
		return
	}
	body, err := json.Marshal(response)
	if err != nil {
		s.writeEarningsMarketFailure(w, "encode market snapshot", err)
		return
	}
	s.readCache.Set(cacheKey, body, 5*time.Minute)
	writeCachedJSON(w, body)
}

func (s *Server) writeEarningsMarketFailure(w http.ResponseWriter, operation string, err error) {
	s.logger.Error("earnings market unavailable", "operation", operation, "error", err)
	writeJSON(w, http.StatusInternalServerError, errorResponse(
		"internal_error",
		"earnings market is temporarily unavailable",
	))
}

func buildEarningsMarketResponse(
	records []store.ModelRegistryRecord,
	aliases []store.ModelAlias,
	capacities []registry.ModelCapacity,
	work []store.ModelSettledWorkTotal,
	windowStart, windowEnd time.Time,
	baseRewardsConfig BaseRewardsConfig,
) (earningsMarketResponse, error) {
	if err := validateEarningsBaseRewardsConfig(baseRewardsConfig); err != nil {
		return earningsMarketResponse{}, err
	}
	capacityByModel := make(map[string]registry.ModelCapacity, len(capacities))
	for _, capacity := range capacities {
		if err := validateEarningsCapacity(capacity); err != nil {
			return earningsMarketResponse{}, err
		}
		if _, duplicate := capacityByModel[capacity.ModelID]; duplicate {
			return earningsMarketResponse{}, fmt.Errorf("duplicate live capacity for model %q", capacity.ModelID)
		}
		capacityByModel[capacity.ModelID] = capacity
	}
	catalog, err := buildEarningsCatalog(records, aliases)
	if err != nil {
		return earningsMarketResponse{}, err
	}

	modelIndex := make(map[string]int, len(catalog))
	models := make([]earningsMarketModel, len(catalog))
	for i, entry := range catalog {
		model := entry.model
		for _, member := range entry.capacityMembers {
			if capacity, ok := capacityByModel[member]; ok {
				model.AggregateTPS += capacity.AggregateTPS
				model.AggregateMemoryBandwidthGBs += capacity.AggregateMemoryBandwidthGBs
				model.BenchmarkTPS += capacity.BenchmarkTPS
				model.BenchmarkMemoryBandwidthGBs += capacity.BenchmarkMemoryBandwidthGBs
				model.ProviderSupply += capacity.EligibleProviders
			}
		}
		models[i] = model
		modelIndex[model.ID] = i
	}

	var audit earningsMarketAudit
	for _, total := range work {
		paidTokens := total.PaidTokens()
		audit.TotalSettledWorkMicroUSD += total.WorkPayoutMicroUSD
		audit.TotalPaidTokens += paidTokens
		audit.TotalPaidJobs += total.Jobs

		index, ok := modelIndex[total.PublicModel]
		if !ok {
			continue
		}
		models[index].WorkPayoutMicroUSD += total.WorkPayoutMicroUSD
		models[index].PaidTokens += paidTokens
		models[index].PaidJobs += total.Jobs
		audit.ModeledWorkMicroUSD += total.WorkPayoutMicroUSD
		audit.ModeledPaidTokens += paidTokens
		audit.ModeledPaidJobs += total.Jobs
	}
	audit.UnattributedWorkMicroUSD = audit.TotalSettledWorkMicroUSD - audit.ModeledWorkMicroUSD
	audit.UnattributedPaidTokens = audit.TotalPaidTokens - audit.ModeledPaidTokens
	audit.UnattributedPaidJobs = audit.TotalPaidJobs - audit.ModeledPaidJobs

	for i := range models {
		models[i].EstimateAvailable, models[i].UnavailableReason = earningsEstimateAvailability(models[i])
	}

	return earningsMarketResponse{
		WindowStart: windowStart,
		WindowEnd:   windowEnd,
		WindowDays:  int(earningsMarketWindow / (24 * time.Hour)),
		Models:      models,
		Audit:       audit,
		BaseRewards: earningsMarketBaseRewards{
			Enabled:             baseRewardsConfig.Enabled,
			MonthlyPoolMicroUSD: baseRewardsConfig.FloorPoolB,
			MinUptimeFraction:   baseRewardsConfig.MinUptimeFrac,
			ReductionK:          baseRewardsConfig.ReductionK,
			AccountCapFraction:  baseRewardsConfig.AccountCapFrac,
			Tiers:               baserewards.Tiers(),
		},
	}, nil
}

func validateEarningsBaseRewardsConfig(config BaseRewardsConfig) error {
	if config.FloorPoolB < 0 ||
		config.MinUptimeFrac < 0 || config.MinUptimeFrac > 1 ||
		math.IsNaN(config.MinUptimeFrac) || math.IsInf(config.MinUptimeFrac, 0) ||
		config.ReductionK < 0 || math.IsNaN(config.ReductionK) || math.IsInf(config.ReductionK, 0) ||
		config.AccountCapFrac < 0 || config.AccountCapFrac > 1 ||
		math.IsNaN(config.AccountCapFrac) || math.IsInf(config.AccountCapFrac, 0) {
		return fmt.Errorf("invalid base reward policy")
	}
	return nil
}

func earningsEstimateAvailability(model earningsMarketModel) (bool, string) {
	if model.WorkPayoutMicroUSD <= 0 || model.PaidTokens <= 0 || model.PaidJobs <= 0 {
		return false, "settled_work_unavailable"
	}
	if model.AggregateTPS <= 0 || model.ProviderSupply <= 0 {
		return false, "competing_capacity_unavailable"
	}
	if model.BenchmarkTPS <= 0 || model.BenchmarkMemoryBandwidthGBs <= 0 {
		return false, "throughput_benchmark_unavailable"
	}
	return true, ""
}

File: console-ui/src/lib/api/earnings-market.ts
import type {
  BaseRewardTier,
  EarningsMarketAudit,
  EarningsMarketBaseRewards,
  EarningsMarketModel,
  EarningsMarketResponse,
  EarningsUnavailableReason,
} from "./types";

type JsonRecord = Record<string, unknown>;

const WINDOW_MS = 30 * 24 * 60 * 60 * 1000;
const INVALID_MARKET_RESPONSE = "Invalid earnings market response";
export const EARNINGS_MARKET_TIMEOUT_MS = 10_000;
const UNAVAILABLE_REASONS = new Set<EarningsUnavailableReason>([
  "settled_work_unavailable",
  "competing_capacity_unavailable",
  "throughput_benchmark_unavailable",
]);

function asRecord(value: unknown): JsonRecord | null {
  return value !== null && typeof value === "object" && !Array.isArray(value)
    ? (value as JsonRecord)
    : null;
}

function isNonNegativeNumber(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value) && value >= 0;
}

function isNonNegativeInteger(value: unknown): value is number {
  return isNonNegativeNumber(value) && Number.isSafeInteger(value);
}

function isPositiveInteger(value: unknown): value is number {
  return isNonNegativeInteger(value) && value > 0;
}

function expectedUnavailableReason(
  model: JsonRecord,
): EarningsUnavailableReason | null {
  if (
    !isNonNegativeInteger(model.work_payout_micro_usd) ||
    !isNonNegativeInteger(model.paid_tokens) ||
    !isNonNegativeInteger(model.paid_jobs) ||
    model.work_payout_micro_usd <= 0 ||
    model.paid_tokens <= 0 ||
    model.paid_jobs <= 0
  ) {
    return "settled_work_unavailable";
  }
  if (
    !isNonNegativeNumber(model.aggregate_tps) ||
    !isNonNegativeInteger(model.provider_supply) ||
    model.aggregate_tps <= 0 ||
    model.provider_supply <= 0
  ) {
    return "competing_capacity_unavailable";
  }
  if (
    !isNonNegativeNumber(model.benchmark_tps) ||
    !isNonNegativeNumber(model.benchmark_memory_bandwidth_gbps) ||
    model.benchmark_tps <= 0 ||
    model.benchmark_memory_bandwidth_gbps <= 0
  ) {
    return "throughput_benchmark_unavailable";
  }
  return null;
}

function isMarketModel(value: unknown): value is EarningsMarketModel {
  const model = asRecord(value);
  if (
    !model ||
    typeof model.id !== "string" ||
    model.id.length === 0 ||
    typeof model.display_name !== "string" ||
    model.display_name.length === 0 ||
    !isPositiveInteger(model.min_ram_gb) ||
    !isPositiveInteger(model.size_bytes) ||
    !isNonNegativeNumber(model.size_gb) ||
    model.size_gb <= 0 ||
    !isNonNegativeInteger(model.work_payout_micro_usd) ||
    !isNonNegativeInteger(model.paid_tokens) ||
    !isNonNegativeInteger(model.paid_jobs) ||
    !isNonNegativeNumber(model.aggregate_tps) ||
    !isNonNegativeNumber(model.aggregate_memory_bandwidth_gbps) ||
    !isNonNegativeNumber(model.benchmark_tps) ||
    !isNonNegativeNumber(model.benchmark_memory_bandwidth_gbps) ||
    !isNonNegativeInteger(model.provider_supply) ||
    typeof model.estimate_available !== "boolean"
  ) {
    return false;
  }
  const expected = expectedUnavailableReason(model);
  if (model.estimate_available) {
    return expected === null && model.unavailable_reason === undefined;
  }
  return (
    expected !== null &&
    typeof model.unavailable_reason === "string" &&
    UNAVAILABLE_REASONS.has(model.unavailable_reason as EarningsUnavailableReason) &&
    model.unavailable_reason === expected
  );
}

function isAudit(value: unknown): value is EarningsMarketAudit {
  const audit = asRecord(value);
  if (!audit) return false;
  if (
    !isNonNegativeInteger(audit.total_settled_work_micro_usd) ||
    !isNonNegativeInteger(audit.modeled_work_micro_usd) ||
    !isNonNegativeInteger(audit.unattributed_work_micro_usd) ||
    !isNonNegativeInteger(audit.total_paid_tokens) ||
    !isNonNegativeInteger(audit.modeled_paid_tokens) ||
    !isNonNegativeInteger(audit.unattributed_paid_tokens) ||
    !isNonNegativeInteger(audit.total_paid_jobs) ||
    !isNonNegativeInteger(audit.modeled_paid_jobs) ||
    !isNonNegativeInteger(audit.unattributed_paid_jobs)
  ) {
    return false;
  }
  const typed = audit as unknown as EarningsMarketAudit;
  return (
    typed.modeled_work_micro_usd + typed.unattributed_work_micro_usd ===
      typed.total_settled_work_micro_usd &&
    typed.modeled_paid_tokens + typed.unattributed_paid_tokens === typed.total_paid_tokens &&
    typed.modeled_paid_jobs + typed.unattributed_paid_jobs === typed.total_paid_jobs
  );
}

function isBaseRewardTier(value: unknown): value is BaseRewardTier {
  const tier = asRecord(value);
  return Boolean(
    tier &&
      isPositiveInteger(tier.min_ram_gb) &&
      isNonNegativeInteger(tier.monthly_micro_usd),
  );
}

function isBaseRewards(value: unknown): value is EarningsMarketBaseRewards {
  const policy = asRecord(value);
  if (
    !policy ||
    typeof policy.enabled !== "boolean" ||
    !isNonNegativeInteger(policy.monthly_pool_micro_usd) ||
    !isNonNegativeNumber(policy.min_uptime_fraction) ||
    policy.min_uptime_fraction > 1 ||
    !isNonNegativeNumber(policy.reduction_k) ||
    !isNonNegativeNumber(policy.account_cap_fraction) ||
    policy.account_cap_fraction > 1 ||
    !Array.isArray(policy.tiers) ||
    policy.tiers.length === 0 ||
    !policy.tiers.every(isBaseRewardTier)
  ) {
    return false;
  }
  const minRAMValues = policy.tiers.map((tier) => tier.min_ram_gb);
  return new Set(minRAMValues).size === minRAMValues.length;
}

export function parseEarningsMarket(value: unknown): EarningsMarketResponse {
  const market = asRecord(value);
  const start = typeof market?.window_start === "string" ? Date.parse(market.window_start) : NaN;
  const end = typeof market?.window_end === "string" ? Date.parse(market.window_end) : NaN;
  if (
    !market ||
    market.window_days !== 30 ||
    !Number.isFinite(start) ||
    !Number.isFinite(end) ||
    end - start !== WINDOW_MS ||
    !Array.isArray(market.models) ||
    market.models.length === 0 ||
    !market.models.every(isMarketModel) ||
    !isAudit(market.audit) ||
    !isBaseRewards(market.base_rewards)
  ) {
    throw new Error(INVALID_MARKET_RESPONSE);
  }

  const models = market.models as EarningsMarketModel[];
  if (new Set(models.map((model) => model.id)).size !== models.length) {
    throw new Error(INVALID_MARKET_RESPONSE);
  }
  const audit = market.audit as unknown as EarningsMarketAudit;
  const modeledWork = models.reduce((sum, model) => sum + model.work_payout_micro_usd, 0);
  const modeledTokens = models.reduce((sum, model) => sum + model.paid_tokens, 0);
  const modeledJobs = models.reduce((sum, model) => sum + model.paid_jobs, 0);
  if (
    modeledWork !== audit.modeled_work_micro_usd ||
    modeledTokens !== audit.modeled_paid_tokens ||
    modeledJobs !== audit.modeled_paid_jobs
  ) {
    throw new Error(INVALID_MARKET_RESPONSE);
  }
  return market as unknown as EarningsMarketResponse;
}

export async function fetchEarningsMarket(): Promise<EarningsMarketResponse> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), EARNINGS_MARKET_TIMEOUT_MS);
  try {
    const response = await fetch("/api/earnings/market", {
      headers: { Accept: "application/json" },
      signal: controller.signal,
    });
    if (!response.ok) {
      throw new Error(`Failed to fetch earnings market: ${response.status}`);
    }
    return parseEarningsMarket(await response.json());
  } finally {
    clearTimeout(timeout);
  }
}
